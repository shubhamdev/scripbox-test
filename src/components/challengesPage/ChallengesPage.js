import React, { useEffect, useState, useRef } from "react";

import { Form, Input, Button, Select, Row, Col, message } from "antd";
import { addToCollection, readCollection } from "../../utils/api";
import { isLogin } from "../../utils";

const { Option } = Select;
const layout = {
  labelCol: { span: 8 },
  wrapperCol: { span: 20 },
};

const validateMessages = {
  required: "${label} is required!",
};

function ChallengesPage(props) {
  const inputRef = useRef(null);
  const [challengeList, setChallengeList] = useState([]);
  const [loginUserName, setLoginUserName] = useState("");
  const [form] = Form.useForm();
  const tagList = [
    "Front end",
    "Back end",
    "Elastic",
    "Nodejs",
    "React",
    "Angular",
  ];

  useEffect(() => {
    if (isLogin()) {
      setLoginUserName(JSON.parse(localStorage.getItem("jwt")));
      inputRef.current.focus();
    }
    readCollection("challenges")
      .then((data) => {
        setChallengeList(data);
      })
      .catch((error) => {});
  }, []);
  const onFinish = ({ challengeData }) => {
    challengeData.createdDate = new Date().toJSON();
    challengeData.createdBy = loginUserName.user.name;
    challengeData.upVote = [];
    debugger;
    addToCollection(challengeData, "challenges")
      .then((data) => {
        const copy = [...challengeList];
        copy.push(challengeData);
        setChallengeList(copy);
        message.success(
          "challenges added successfully" + " " + "ID: " + data.id
        );
        form.resetFields();
      })
      .catch((error) => {
        message.error(
          "Something went wrong please try again." + " " + "Error: " + error
        );
        form.resetFields();
      });
  };

  return (
    <div className="site-card-wrapper">
      <Row style={{ padding: "5%" }}>
        <Col offset={0} span={20}>
          <Form
            {...layout}
            form={form}
            name="challenge-from"
            onFinish={onFinish}
            validateMessages={validateMessages}
          >
            <Form.Item
              name={["challengeData", "title"]}
              label="Title"
              rules={[
                {
                  required: true,
                  message: "Please confirm your password!",
                },
                {
                  validator: (_, value) => {
                    const data = challengeList.filter((item) => {
                      if (item.title.toLowerCase() === value.toLowerCase()) {
                        return item;
                      }
                    });
                    if (data && data.length === 0) {
                      return Promise.resolve();
                    }
                    return Promise.reject(
                      new Error(
                        "This title is already present please try another title."
                      )
                    );
                  },
                },
              ]}
            >
              <Input
                ref={inputRef}
                autoComplete="off"
                placeholder="Enter title"
              />
            </Form.Item>
            <Form.Item
              name={["challengeData", "tag"]}
              // name="gender"
              label="Tag"
              rules={[{ required: true }]}
            >
              <Select
                placeholder="Select a tag"
                // onChange={onGenderChange}
                allowClear
              >
                {tagList.map((item) => (
                  <Option value={item} key={item}>
                    {item}
                  </Option>
                ))}
              </Select>
            </Form.Item>

            <Form.Item
              name={["challengeData", "description"]}
              rules={[{ required: true }]}
              label="Description"
            >
              <Input.TextArea
                autoComplete="off"
                placeholder="Enter description"
              />
            </Form.Item>
            <Form.Item wrapperCol={{ ...layout.wrapperCol, offset: 22 }}>
              <Button type="primary" htmlType="submit">
                Submit
              </Button>
            </Form.Item>
          </Form>
        </Col>
      </Row>
    </div>
  );
}

export default ChallengesPage;
