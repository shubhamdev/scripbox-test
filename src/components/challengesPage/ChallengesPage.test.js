import React from "react";
import { unmountComponentAtNode } from "react-dom";
import { shallow } from "enzyme";
import ChallengesPage from "./ChallengesPage";

let container = null;

beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe("ChallengesPage component", () => {
  it('should render ChallengesPage component correctly in "debug" mode', () => {
    const component = shallow(<ChallengesPage debug />);
    expect(component).toMatchSnapshot();
  });
  it("should render ChallengesPage component correctly with no props", () => {
    const component = shallow(<ChallengesPage />);
    expect(component).toMatchSnapshot();
  });
});
