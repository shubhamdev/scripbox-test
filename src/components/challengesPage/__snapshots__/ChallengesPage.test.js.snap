// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChallengesPage component should render ChallengesPage component correctly in "debug" mode 1`] = `
<div
  className="site-card-wrapper"
>
  <Row
    style={
      Object {
        "padding": "5%",
      }
    }
  >
    <Col
      offset={0}
      span={20}
    >
      <ForwardRef(InternalForm)
        form={
          Object {
            "__INTERNAL__": Object {
              "itemRef": [Function],
            },
            "getFieldError": [Function],
            "getFieldInstance": [Function],
            "getFieldValue": [Function],
            "getFieldsError": [Function],
            "getFieldsValue": [Function],
            "getInternalHooks": [Function],
            "isFieldTouched": [Function],
            "isFieldValidating": [Function],
            "isFieldsTouched": [Function],
            "isFieldsValidating": [Function],
            "resetFields": [Function],
            "scrollToField": [Function],
            "setFields": [Function],
            "setFieldsValue": [Function],
            "submit": [Function],
            "validateFields": [Function],
          }
        }
        labelCol={
          Object {
            "span": 8,
          }
        }
        name="challenge-from"
        onFinish={[Function]}
        validateMessages={
          Object {
            "required": "\${label} is required!",
          }
        }
        wrapperCol={
          Object {
            "span": 20,
          }
        }
      >
        <FormItem
          label="Title"
          name={
            Array [
              "challengeData",
              "title",
            ]
          }
          rules={
            Array [
              Object {
                "message": "Please confirm your password!",
                "required": true,
              },
              Object {
                "validator": [Function],
              },
            ]
          }
        >
          <Input
            autoComplete="off"
            placeholder="Enter title"
            type="text"
          />
        </FormItem>
        <FormItem
          label="Tag"
          name={
            Array [
              "challengeData",
              "tag",
            ]
          }
          rules={
            Array [
              Object {
                "required": true,
              },
            ]
          }
        >
          <ForwardRef(InternalSelect)
            allowClear={true}
            placeholder="Select a tag"
          >
            <Option
              key="Front end"
              value="Front end"
            >
              Front end
            </Option>
            <Option
              key="Back end"
              value="Back end"
            >
              Back end
            </Option>
            <Option
              key="Elastic"
              value="Elastic"
            >
              Elastic
            </Option>
            <Option
              key="Nodejs"
              value="Nodejs"
            >
              Nodejs
            </Option>
            <Option
              key="React"
              value="React"
            >
              React
            </Option>
            <Option
              key="Angular"
              value="Angular"
            >
              Angular
            </Option>
          </ForwardRef(InternalSelect)>
        </FormItem>
        <FormItem
          label="Description"
          name={
            Array [
              "challengeData",
              "description",
            ]
          }
          rules={
            Array [
              Object {
                "required": true,
              },
            ]
          }
        >
          <ForwardRef
            autoComplete="off"
            placeholder="Enter description"
          />
        </FormItem>
        <FormItem
          wrapperCol={
            Object {
              "offset": 22,
              "span": 20,
            }
          }
        >
          <Button
            htmlType="submit"
            type="primary"
          >
            Submit
          </Button>
        </FormItem>
      </ForwardRef(InternalForm)>
    </Col>
  </Row>
</div>
`;

exports[`ChallengesPage component should render ChallengesPage component correctly with no props 1`] = `
<div
  className="site-card-wrapper"
>
  <Row
    style={
      Object {
        "padding": "5%",
      }
    }
  >
    <Col
      offset={0}
      span={20}
    >
      <ForwardRef(InternalForm)
        form={
          Object {
            "__INTERNAL__": Object {
              "itemRef": [Function],
            },
            "getFieldError": [Function],
            "getFieldInstance": [Function],
            "getFieldValue": [Function],
            "getFieldsError": [Function],
            "getFieldsValue": [Function],
            "getInternalHooks": [Function],
            "isFieldTouched": [Function],
            "isFieldValidating": [Function],
            "isFieldsTouched": [Function],
            "isFieldsValidating": [Function],
            "resetFields": [Function],
            "scrollToField": [Function],
            "setFields": [Function],
            "setFieldsValue": [Function],
            "submit": [Function],
            "validateFields": [Function],
          }
        }
        labelCol={
          Object {
            "span": 8,
          }
        }
        name="challenge-from"
        onFinish={[Function]}
        validateMessages={
          Object {
            "required": "\${label} is required!",
          }
        }
        wrapperCol={
          Object {
            "span": 20,
          }
        }
      >
        <FormItem
          label="Title"
          name={
            Array [
              "challengeData",
              "title",
            ]
          }
          rules={
            Array [
              Object {
                "message": "Please confirm your password!",
                "required": true,
              },
              Object {
                "validator": [Function],
              },
            ]
          }
        >
          <Input
            autoComplete="off"
            placeholder="Enter title"
            type="text"
          />
        </FormItem>
        <FormItem
          label="Tag"
          name={
            Array [
              "challengeData",
              "tag",
            ]
          }
          rules={
            Array [
              Object {
                "required": true,
              },
            ]
          }
        >
          <ForwardRef(InternalSelect)
            allowClear={true}
            placeholder="Select a tag"
          >
            <Option
              key="Front end"
              value="Front end"
            >
              Front end
            </Option>
            <Option
              key="Back end"
              value="Back end"
            >
              Back end
            </Option>
            <Option
              key="Elastic"
              value="Elastic"
            >
              Elastic
            </Option>
            <Option
              key="Nodejs"
              value="Nodejs"
            >
              Nodejs
            </Option>
            <Option
              key="React"
              value="React"
            >
              React
            </Option>
            <Option
              key="Angular"
              value="Angular"
            >
              Angular
            </Option>
          </ForwardRef(InternalSelect)>
        </FormItem>
        <FormItem
          label="Description"
          name={
            Array [
              "challengeData",
              "description",
            ]
          }
          rules={
            Array [
              Object {
                "required": true,
              },
            ]
          }
        >
          <ForwardRef
            autoComplete="off"
            placeholder="Enter description"
          />
        </FormItem>
        <FormItem
          wrapperCol={
            Object {
              "offset": 22,
              "span": 20,
            }
          }
        >
          <Button
            htmlType="submit"
            type="primary"
          >
            Submit
          </Button>
        </FormItem>
      </ForwardRef(InternalForm)>
    </Col>
  </Row>
</div>
`;
